server.c

```diff
@@ -147,6 +151,8 @@ void handle_solution(int fd, struct msg_solution *solution)
         task_generate(current_block);
         LOG("Generated new block: %s\n", current_block);
     }
+
+    pthread_mutex_unlock(&lock);
 }
```

Comment:
We can rework this slightly: lock before verification. After verification, generate new task if necessary, then unlock. After unlocking we can write the reply back to the client. This way we spend much less time locked + blocking other threads, especially during a blocking call to `write`